pkg/controllers/jobset_controller.go

```diff
@@ -142,28 +143,62 @@ func (r *JobSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Complete(r)
 }
 
-func (r *JobSetReconciler) constructJobFromTemplate(js *jobset.JobSet, rjob *jobset.ReplicatedJob) (*batchv1.Job, error) {
-	job := &batchv1.Job{
-		ObjectMeta: metav1.ObjectMeta{
-			Labels:      make(map[string]string),
-			Annotations: make(map[string]string),
-			Name:        genJobName(js, rjob),
-			Namespace:   js.Namespace,
-		},
-		Spec: *rjob.Template.Spec.DeepCopy(),
-	}
+func (r *JobSetReconciler) constructJobsFromTemplate(js *jobset.JobSet, rjob *jobset.ReplicatedJob, ownedJobs *childJobs) ([]*batchv1.Job, error) {
+	var jobs []*batchv1.Job
 
-	// If enableDNSHostnames is set, update job spec to set subdomain as
-	// job name (a headless service with same name as job will be created later).
-	if rjob.Network.EnableDNSHostnames != nil && *rjob.Network.EnableDNSHostnames {
-		job.Spec.Template.Spec.Subdomain = job.Name
+	// Defaulting and validation.
+	// TODO: Do defaulting and validation in webhook instead of here (https://github.com/kubernetes-sigs/jobset/issues/6)
+	replicas := 1
+	if rjob.Replicas != nil && *rjob.Replicas > 0 {
+		replicas = *rjob.DeepCopy().Replicas
 	}
 
-	// Set controller owner reference for garbage collection and reconcilation.
-	if err := ctrl.SetControllerReference(js, job, r.Scheme); err != nil {
-		return nil, err
+	// Construct jobs.
+	for i := 0; i < replicas; i++ {
+		// Check if we need to create this job. If not, skip it.
+		jobName := genJobName(js, rjob, i)
+		if create := r.shouldCreateJob(jobName, ownedJobs); !create {
+			continue
+		}
+
+		// Copy labels/annotations to avoid modifying the template itself.
+		labels := rjob.Template.Labels
```

Comment:
This is just copying the reference, we need to do a proper copy:

```
labels = make(map[string]string)
for k, v := rjob.Template.Labels {
  labels[k] = v
}
```

Reply:
Woops, can't believe I overlooked that. Fixed!